keps/sig-apps/3715-elastic-indexed-job/README.md

```diff
@@ -0,0 +1,936 @@
+- Maintain the same success, failure and completion semantics of `Indexed` mode.
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+### Non-Goals
+
+- Change the existing behavior of `NonIndexed` and `Indexed` modes.
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+I have a batch workload that requires two sets of pods: a manager and a group of workers that can scale up or down. 
+The manager communicates with the workers to manage their load; examples of such workloads include distributed HPC 
+and districuted ML training. Examples of frameworks include Horovod with MPI, Spark and Ray.
+
+This workload can be modeled as two Jobs, an `Indexed` Job for the manager and an `ElasticIndexed` Job 
+for the workers. A headless service is created to set up stable hostnames for the worker pods . The workers job is
+scaled up/down by updating `spec.completions` and `spec.parallelism` in tandem.
+
+
+<!--
+
+### Notes/Constraints/Caveats (Optional)
+
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+There is a concern that changing completion semantics on the fly can be subject to
+race conditions with declaring the job finished. However, after a closer look, this 
+seems to not be an issue because the job controller uses Update when setting the
+finished status on the job object, which is subject to object modified errors if the spec
+changes at the same time.
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+
+On the job controller side, the only required change is to ensure that `ElasticIndexed` is treated as an `Indexed` Job. 
```

Comment:
If we're going with a new completion mode, why do we want to speed up the implementation, what's wrong with the usual alpha->beta->ga process?

Reply:
This is not accelerating, this is required to safely add a new enum

Reply:
I'd consider new enum, as a new feature, thus I'm not sure I understand the need for not going the usual path?

Reply:
This is not an unusual path, see https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#new-enum-value-in-existing-field

but in any case, lets first converge on whether or not we want to generally allow to mutate completions, if so, we will not need a new enum at all.